### PR TITLE
Out of memory caching

### PR DESCRIPTION
**Pull request by jimhester**

### CODE REVIEW

File: DESCRIPTION [hunk 1]
```diff
@@ -1,18 +1,26 @@
...
+    person("Daniel", "Cook", , "danielecook@gmail.com", role = "aut"))
 Description: Cache the results of a function so that when you call it
     again with the same arguments it returns the pre-computed value.
-URL: https://github.com/hadley/memoise
+URL: https://github.com/danielecook/xmemoise
 BugReports: https://github.com/hadley/memoise/issues
 Imports:
-    digest (>= 0.6.3)
+    digest (>= 0.6.3),
+    googleAuthR,
```
Comment (jimhester) [inline DESCRIPTION]: cache specific dependencies should probably go in `Suggests:` instead of `Imports:`

File: DESCRIPTION [hunk 2]
```diff
@@ -1,18 +1,26 @@
...
+    base64enc
 Suggests:
     testthat
+Remotes:
+    cloudyr/aws.s3
```
Comment (jimhester) [inline DESCRIPTION]: We have to be somewhat careful about dependencies not yet on CRAN, `aws.s3` also should be in `Suggests:` as well as `Remotes:`.

File: NAMESPACE [hunk 3]
```diff
@@ -1,12 +1,15 @@
...
 S3method(print,memoised)
+export(cache_datastore)
+export(cache_s3)
 export(forget)
 export(has_cache)
 export(is.memoised)
 export(is.memoized)
 export(memoise)
 export(memoize)
+export(new_cache)
```
Comment (jimhester) [inline NAMESPACE]: The existing behavior should be renamed to `cache_local`

File: R/cache_datastore.r [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+  query_ds <- googleAuthR::gar_api_generator(paste0(base_url, ":runQuery"),
+                                             "POST",
+                                             data_parse_function = function(resp) resp)
+
+
+  cache_reset <- function() {
+    query_results <- query_ds(the_body = list(gqlQuery = list(queryString = paste0("SELECT * FROM ", cache_name))))
+    while((query_results$batch$moreResults != "NO_MORE_RESULTS") | is.null(query_results$batch$entityResults) == FALSE) {
+        ids <- (query_results$batch$entityResults$entity$key$path %>% dplyr::bind_rows())$name
```
Comment (jimhester) [inline R/cache_datastore.r]: You will need to remove use of `%>%` and the dplyr dependency here.

I don't know exactly what data structure is being returned, but from the code this looks like you can use `vapply()` with an extractor function.

**Comment by jimhester** [comment]
@danielecook Please look at the travis build results (https://travis-ci.org/hadley/memoise/builds/148021731#L1180-L1217) and fix all Warnings, most are due to documentation mismatches or unimported functions.

**Comment by danielecook** [comment]
@jimhester Hey thanks! I wasn't sure whether the authors would want to add this so I renamed the package and had started my own fork but I'll change it back and resolve these issues. I'll also add a file-based cache soon (for dropbox/Google Drive). I was in a bit of a rush to get started using this for some of my own projects (-;

**Comment by danielecook** [comment]
Got this to work, but had to remove quite a bit of documentation. I tend not to be so stringent as to not allow things like:
```
cache_filesystem("~/Dropbox/.rcache")
```

**Comment by jimhester** [comment]
They should be in `#' @examples` not `#' @usage`, and if you don't want them to run you need to put them in a `dontrun` block.
```r
#' \dontrun{
#'  # ... code to not run
#' }
```

See (http://r-pkgs.had.co.nz/man.html#man-functions) for a good reference.

**Comment by jimhester** [comment]
I ended up merging this with https://github.com/hadley/memoise/commit/b3b9ae66f5798c6d6ac861e77ff17d515588dd72.

I cleaned up some things, but decided to remove the google datastore backend, there was too much API specific code that I don't think belongs in this package. Also I was a little confused why you were using datastore for this, Google cloud store (with https://github.com/cloudyr/googleCloudStorageR) seems like a better fit.

Thank you again for the PR and sorry it took so long to get merged. I think this is a very nice improvement to the package!

**Comment by danielecook** [comment]
Hey just saw these options were folded in. Thanks for taking care of this! Google Cloud storage is a better fit but it's also probably slower. Regardless, glad to see these changes implemented. Have a good rest of the year!

**Pull request closed without merging**